### ISSUE TITLE
Null is very non-intuitive as a way to say "i don't care" when subclassing a generic type

### ISSUE THREAD

**Issue by Hixie** [dart-lang/sdk#34416]
Sometimes, there is a class and it has a generic type, and it has some subclasses that work for all types. For example:

```dart
abstract class MenuItem<T> {
  String get label;
  bool matches(T value);
}
     
class Option<T> extends MenuItem<T> {
  Option(this.value);
  final T value;
  String get label => '$value';
  bool matches(T value) => value == this.value;
}
             
class Divider extends MenuItem<T> { // but what should T be?                                                                            
  String get label => '---';
  bool matches(T value) => false; // and what about here?
}
```

For example, this happens with Flutter's `PopupMenuEntry<T>` class hierarchy. I've also seen it with other cases, for example, I have one project where there's a `Progress<T>` class which can report progress on getting a value of type `T`, and there's a subclass that represents having hit an error, and that class doesn't care about `T`.

Currently, the only way to do this is to use `Null` as the value of `T` for the "general" subclasses. However, this is _super_ unintuitive. In Dart 1, you used to be able to use "dynamic", which is better, though still not ideal. It would be great if there was a way to do this that was clearer and more obvious.

**Comment by a14n** [comment]
Couldn't be `void` used here?

**Comment by Hixie** [comment]
I don't think that's particularly intuitive either. :-)

**Comment by lrhn** [comment]
This is a classical example - a divider *is not* a menu item  (it's not, really - you can't select it, you need to skip it when you use arrow navigation, it cannot have a key shortcut key, it doesn't really have a label, it's a square peg in a round hole), but it has to co-exist with menu items in the list of menu content.

Here the first step has been taken, by separating `Option` from `Divider`, but there are still members of the `MenuItem` superclass that do not apply to `Divider`.

The honest modeling would be that a `Menu` contains `MenuContent`, and `Divider` and `MenuItem` are separate `MenuContent` classes. The `Menu` should probably allow you to iterate its `items` without seeing the dividers.

(Another alternative is to have separate groups of items, and automatically display dividers between them, rather than having to add dividers - logical formatting vs physical formatting).

With the current model, I'd use `Null` as the element type. It's a `MenuItem<T>` for any `T`, so it fits in all menues, and it clearly signifies that the value of the "item" is irrelevant. You can't use `void` (it's not a `MenuItem<String>` so it won't fit in a `Menu<String>`).

I do think that `MenuItem<Null>` very clearly states that there is no interesting value here.

I also think that it generalizes to other cases where some class works at all types - the only value which does work at all types is `null`, so that must be the type. If we get non-nullable types, we'll likely need a bottom type below those, exactly to handle these cases. That's a type that contains no value (but sadly won't be called `void`, since that one is taken).

**Comment by eernstg** [comment]
Apart from the concrete class hierarchy design, there's the issue of considering `Null` as a natural choice when a "general" entity is needed. I know that it isn't easy to argue convincingly that any particular solution is intuitive, because that's a property of a person's state of mind. But here's a point of view that might help:

To create a **"general type annotation"**, e.g., in order to declare a variable whose value can be drawn from the greatest possible set of instances, use a **top type**. So a general variable in the extreme case could have type `Object`, and a "general list" variable could have type `List<Object>`, etc. You may opt in to a more forgiving static analysis by using `dynamic`, but that's a separate issue.

Conversely, to create a **"general instance"**, e.g., in order to create a menu item that can be passed around anywhere a menu item is expected (and used in a read-only manner), use a **bottom type**, that is, `Null`.

There's even a spatial metaphor for this: In order to be a container that is able to receive and hold anything, be big; in order to be able to go into any container, be small.

**Comment by Hixie** [comment]
As I was laying in bed last night I concluded that actually `void` would convey the right thing — it's saying "I won't read this value" (e.g. in the `matches` method above).

`Null` isn't something I would have guessed. I spent quite a long time trying to figure out what type to use here, last night. I tried `dynamic`, and after that was a bit stumped, until I remembered that Flutter does this for menu items, and that's when I found we had used `Null`.

`I'm open to alternative approaches. I'm not sure I understand the `MenuContent` idea. The desire is to provide a tree that represents what we should render, with static analysis catching the case of `MenuItem<T>` being used with the wrong `T` for the menu (each menu has a particular type of value that it can return).

**Comment by MichaelRFairhurst** [comment]
I have a different thought. Rather than complaining about `Null`, I'd like to point out that `MenuItem` is a contravariant class. `MenuItem<void>` (or `MenuItem<dynamic>`, too) is the correct type in a world where `MenuItem<T>` can be, and is, marked contravariant.

To be fair, I would already like to see a type on top of Null. It would just be there as a stricter set of Null -- sure, null is allowable anywhere, so a Bottom type that allows null makes sense. But a Bottom type that rejects Null has value too.\*

I also think it would get past the confusion a bit here -- Null looks like a type that should be easy to comprehend and intuitive, and its anything but.

The difficulty in naming the new Bottom, is that both top and bottom are both everything and nothing:

* Object is not provably anything, because it may be everything
* Bottom is assumably everything, because it must be nothing.

I think if they were named `Anything`/`Nothing` this would be pretty clear...though there would be confusion over `void`/`Nothing`. 

The current state of `void`/`Object`/`dynamic`/`Null` is already confusing enough. I think `void`/`Object`/`dynamic`/`Nothing` would be an improvement....but we have no reason to _remove_ `Null`, which probably means `void`/`Object`/`dynamic`/`Null`/`Nothing` is even worse. (we could lint `Null` much more effectively in this world though, with a rule adapted from `prefer_void_to_null`).

Bikeshedding out-of-the-box names for `Nothing`: `Assumed`, `Empty`, `Placeholder`, `Everything`...

Really the best description of Bottom is: "the empty set of values & the subtype of all types," but there isn't exactly a good english word that captures that...

\* `() -> Null` may return a value, and `() -> Bottom` would not be able to. This has fringe benefits...one example is accepting a `Bottom Function(Error) handler` which is guaranteed to throw (or at least, not resume execution in the caller).

**Comment by eernstg** [comment]
> .. a Bottom type that rejects Null

Right, I think that we'll need a `Nothing` type denoting the empty set when we add non-null types, because otherwise we won't be able to specify those "general instances" or, for contravariant placement "all-inclusive function types":

```dart
// Assuming types are non-null by default, nullable with `?`.

Nothing fn() => throw 'Never completes with a value';

int? Function() f1 = fn; // OK
int Function() f2 = fn; // OK, but it couldn't be a `Null Function()`.

List<Nothing> ln = [];

List<int> ints = ln; // OK
// etc.
```

I'm not quite sure that `Nothing` would take over completely as the only relevant bottom type. It is also possible that `Null` can be useful &mdash; we're now talking about situations where the type specifies that `null` is allowed because someone really wanted that and wrote an explicit `?`, and in that situation a `null` could be delivered and play the role as "something which is missing", or whatever `null` is taken to mean.

With that, I think it's quite meaningful to name the class that contains `null` (and nothing else) as `Null`. YMMV, of course. ;-)

**Comment by Hixie** [comment]
One problem with `Nothing` is that it's novel. I'm not sure it's something I've seen in any language that's commonly used, and its behaviour and semantics are not intuitively derivable from experience with either Dart or other commonly used languages.

Also, I think we should be wary of any solution that requires an understanding of "top" vs "bottom" to explain, because very few of our users are going to be comfortable enough with types that they will be able to follow such an explanation.

**Comment by MichaelRFairhurst** [comment]
> Right, I think that we'll need a Nothing type denoting the empty set when we add non-null types

I'm going beyond "will need," I'm saying "already could use."

For instance, `Null` is confusing, and if we had `Never`, then we could easily lint usages of `Null`, and over time the language would become less confusing. (maybe).

It also solves the problem I showed before where you can accept a `Never Function(Error) handler`, and you have some guarantee that users won't pass in, for instance, `(e) { print(e); }`, causing issues when:

```
  if (!file.exists) {
    handler(Error("won't delete folder, it doesn't appear to be a build output directory."));
  }
  // oops, handler can return Null instead of throw/exit/loop!
  folder.delete();
```

And it makes it easier to have a conversation about it because you don't have to convince people that `Null` isn't `null` _before even getting to the hard part_.

**Comment by Hixie** [comment]
> The bottom type exists in Typescript, Rust, Scala, and (kinda) C++. 

Yeah, that's fair. The Typescript case is particularly compelling. (Rust is an interesting case in that they explicitly call out the `!` type as not being bottom.)

**Comment by MichaelRFairhurst** [comment]
Ah interesting re: rust. I didn't realize the RFC for Bang as the Bottom type wasn't submitted (I thought that was the RFC _introducing_ bang, so I didn't read it :))

**Comment by MichaelRFairhurst** [comment]
Can we close this to make two issues:

1. a true Bottom type "Never"
2. contravariant generics

?

**Comment by Hixie** [comment]
@mofada Can you file a bug on Flutter with this? That seems serious. This bug was about it being non-intuitive, not it being entirely broken.

**Comment by mofada** [comment]
@Hixie Okay thank you

**Comment by mofada** [comment]
@Hixie I have already submitted [PopupMenuButton attribute initialValue with itemBuilder add PopupMenuDivider](https://github.com/flutter/flutter/issues/27072), thank you
